PR: https://github.com/odoo/odoo/pull/111505

From: 901e3814fc1f15f62c665f0a1a35fabb6dff5e4c
From: Kevin Baptiste
Date: 2023-02-14 14:52:03

Structural Changes: 7
Total Changes: 33

[IMP] hr: allow users to manage plan activities

When starting an onboarding/offboarding plan on an employee, activities
were only created for the users that were part of the HR Officer group.

Regular users can't access the `hr.employee` model and it's not possible
to have activities on `hr.employee.public`.

A new model has been created to manage those activities and is only
accessible to the user and their manager.

task-3151758

closes odoo/odoo#111505

Signed-off-by: Kevin Baptiste <kba@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr/models/__init__.py
+++ b/addons/hr/models/__init__.py
@@ -11,6 +11,7 @@ from . import hr_departure_reason
 from . import hr_job
 from . import hr_plan
 from . import hr_plan_activity_type
+from . import hr_plan_employee_activity
 from . import hr_work_location
 from . import mail_alias
 from . import mail_channel

--- a/None
+++ b/addons/hr/models/hr_plan_employee_activity.py
@@ -0,0 +1,32 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+
+class HrPlanEmployeeActivity(models.Model):
+    _name = 'hr.plan.employee.activity'
+    _description = 'Plan'
+    _inherit = ['mail.thread', 'mail.activity.mixin']
+    _mail_post_access = 'read'
+    _rec_name = 'employee_id'
+
+    employee_id = fields.Many2one('hr.employee', required=True, ondelete='cascade')
+    department_id = fields.Many2one(related='employee_id.department_id')
+    image_128 = fields.Image(related='employee_id.image_128')
+    company_id = fields.Many2one(related='employee_id.company_id')
+    user_id = fields.Many2one(related='employee_id.user_id')
+    summary = fields.Html(compute='_compute_summary')
+
+    @api.depends('activity_ids.summary')
+    def _compute_summary(self):
+        for plan in self:
+            plan.summary = self.env['ir.ui.view']._render_template('hr.hr_employee_plan_activity_summary', {
+                'activity_ids': plan.activity_ids,
+            })
+
+    @api.autovacuum
+    def _gc_employee_plan_activity(self):
+        no_activities = self.env['hr.plan.employee.activity'].search(
+            [('activity_ids', '=', False)]
+        )
+        no_activities.unlink()
